keras/layers/wrappers.py

```diff
@@ -152,9 +157,33 @@ def compute_output_shape(self, input_shape):
         timesteps = input_shape[1]
         return (child_output_shape[0], timesteps) + child_output_shape[1:]
 
+    def _reshape_input_mask(self, mask):
+        if mask is not None and K.ndim(mask) > 2:
+            mask_shape = K.shape(mask)
+            mask_shape = tuple(mask_shape[i] for i in range(K.ndim(mask)))
+            return K.reshape(mask, (-1,) + mask_shape[2:])
+        return None
+
+    def compute_mask(self, inputs, mask):
+        input_shape = K.int_shape(inputs)
+        input_length = input_shape[1]
+        if input_length is None:
+            input_length = K.shape(inputs)[1]
+        inputs = K.reshape(inputs, (-1,) + input_shape[2:])
+        reshaped_mask = self._reshape_input_mask(mask)
+        output_mask = self.layer.compute_mask(inputs, reshaped_mask)
```

Comment:
This mechanism will only work for a few layers, and will fail in the general case. The proper behavior when batch size matters is to slice the input mask and run slices through `self.layer.compute_mask`, then concatenate. No reshaping.